plan.go

```diff
@@ -0,0 +1,92 @@
+package tfe
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"time"
+)
+
+// PlanStatus represents a plan state.
+type PlanStatus string
+
+//List all available plan statuses.
+const (
+	PlanCanceled   PlanStatus = "canceled"
+	PlanCreated    PlanStatus = "created"
+	PlanErrored    PlanStatus = "errored"
+	PlanFinished   PlanStatus = "finished"
+	PlanMFAWaiting PlanStatus = "mfa_waiting"
+	PlanPending    PlanStatus = "pending"
+	PlanQueued     PlanStatus = "queued"
+	PlanRunning    PlanStatus = "running"
+)
+
+// Plans handles communication with the plan related methods of the Terraform
+// Enterprise API.
+//
+// TFE API docs: https://www.terraform.io/docs/enterprise/api/plan.html
+type Plans struct {
+	client *Client
+}
+
+// Plan represents a Terraform Enterprise plan.
+type Plan struct {
+	ID               string                `jsonapi:"primary,plans"`
+	HasChanges       bool                  `jsonapi:"attr,has-changes"`
+	LogReadURL       string                `jsonapi:"attr,log-read-url"`
+	Status           PlanStatus            `jsonapi:"attr,status"`
+	StatusTimestamps *PlanStatusTimestamps `jsonapi:"attr,status-timestamps"`
+}
+
+// PlanStatusTimestamps holds the timestamps for individual plan statuses.
+type PlanStatusTimestamps struct {
+	CanceledAt   time.Time `json:"canceled-at"`
+	CreatedAt    time.Time `json:"created-at"`
+	ErroredAt    time.Time `json:"errored-at"`
+	FinishedAt   time.Time `json:"finished-at"`
+	MFAWaitingAt time.Time `json:"mfa_waiting-at"`
+	PendingAt    time.Time `json:"pending-at"`
+	QueuedAt     time.Time `json:"queued-at"`
+	RunningAt    time.Time `json:"running-at"`
+}
+
+// Read a single run by its ID.
+func (s *Plans) Read(ctx context.Context, planID string) (*Plan, error) {
+	if !validStringID(&planID) {
+		return nil, errors.New("Invalid value for plan ID")
+	}
+
+	u := fmt.Sprintf("plans/%s", url.QueryEscape(planID))
+	req, err := s.client.newRequest("GET", u, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	p := &Plan{}
+	err = s.client.do(ctx, req, p)
+	if err != nil {
+		return nil, err
+	}
+
+	return p, nil
+}
+
+// Logs retrieves the (partial) logs of a plan.
+func (s *Plans) Logs(ctx context.Context, url string) ([]byte, error) {
+	req, err := s.client.newRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Accept", "text/plain")
+
+	var buf bytes.Buffer
+	err = s.client.do(ctx, req, &buf)
```

Comment:
I think it would be nice if we could return an `io.Reader` here so that the logs could be streamed in. Some plan logs can be very verbose depending on if the user enabled debug or trace logs, and the number of resources in the config. Also wondering about log streaming. It could be nice to use the `io.Reader` to implement this (not necessarily in this PR of course), because we could basically just return the reader, and the user could continue to call `Read()` on it until they get an EOF (which we would return when they reached the end of the log).

Reply:
I played with a couple of options, but it doesn't make the code and using the API easier. The main reason for that is that currently the API _doesn't_ stream the logs. So even if we return an `io.Reader`, you will still need to `ioutil.ReadAll()` to get all the bytes and then trim the bytes that you already showed in the previous iteration (think about "streaming" the logs to the console here).

The best I could come up with is this one:

```go
// Logs retrieves the (partial) logs of a plan.
func (p *Plan) Logs(ctx context.Context) (io.Reader, error) {
	// Create an io.Pipe for piping the logs to a reader.
	r, w := io.Pipe()

	// Close the writer and return if there isn't a log URL.
	if p.LogReadURL == "" {
		w.Close()
		return r, nil
	}

	// Create the request.
	req, err := http.NewRequest("GET", p.LogReadURL, nil)
	if err != nil {
		return nil, err
	}
	req = req.WithContext(ctx)

	// Retrieve the available logs.
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}

	// Basic response checking.
	if err := checkResponseCode(resp); err != nil {
		resp.Body.Close()
		return nil, err
	}

	go func() {
		defer w.Close()
		defer resp.Body.Close()
		io.Copy(w, resp.Body)
	}()

	return r, nil
}
```

This looks pretty OK, but with the downside I mentioned before. The application using this will need to diff (trim) the logs before knowing what to display. This makes the reader a bit useless.

So I'm thinking lets keep and merge the code as is for this API call. And then when we get a real streaming endpoint, we can keep this endpoint for retrieving the complete logs and add new `Logs()` method on the `Plan` struct that will use the new streaming capabilities.

Reply:
We actually do already have a streaming log API! It's what we use in the TFE web interface. You request the log in chunks, and track the last byte index you have seen. That allows you to iterate through the log without re-requesting the whole thing every time. See the streaming options [here](https://github.com/hashicorp/archivist#get-v1objectobject). My thinking was that you could still just call `tfe.Plans.Logs(ctx, plan)`, and then just read until `EOF` to get the full log out. Internally If you use something like `io.Copy` this would already be handled for you, making it extremely easy to stream the log. We don't have to do the full streaming API immediately but I think returning the whole log as an in-memory byte slice is not quite right for a relatively unpredictable log size.

Reply:
> We actually do already have a streaming log API!

Ah... That I didn't know! That changes things of course :) Let me take a look at it next week, as I like to do it "first time right". So I prefer to then go ahead and move the `Logs()` method to the `Plan` struct and use the streaming API.

I'll check it out and make an update on the PR 👍